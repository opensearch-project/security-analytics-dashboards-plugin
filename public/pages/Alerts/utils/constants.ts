/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { ALERT_STATE } from '../../../utils/constants';

export const DISABLE_ACKNOWLEDGED_ALERT_HELP_TEXT = 'Only active alerts can be acknowledged.';

export enum AlertSeverity {
  ONE = '1',
  TWO = '2',
  THREE = '3',
  FOUR = '4',
  FIVE = '5',
}

export const severityOptions = [
  { value: 'ALL', text: 'All severity levels' },
  { value: AlertSeverity.ONE, text: '1 (Highest)' },
  { value: AlertSeverity.TWO, text: '2 (High)' },
  { value: AlertSeverity.THREE, text: '3 (Medium)' },
  { value: AlertSeverity.FOUR, text: '4 (Low)' },
  { value: AlertSeverity.FIVE, text: '5 (Lowest)' },
];

export const stateOptions = [
  { value: 'ALL', text: 'All alerts' },
  { value: ALERT_STATE.ACTIVE, text: 'Active' },
  { value: ALERT_STATE.ACKNOWLEDGED, text: 'Acknowledged' },
  { value: ALERT_STATE.COMPLETED, text: 'Completed' },
  { value: ALERT_STATE.ERROR, text: 'Error' },
  { value: ALERT_STATE.DELETED, text: 'Deleted' },
];
