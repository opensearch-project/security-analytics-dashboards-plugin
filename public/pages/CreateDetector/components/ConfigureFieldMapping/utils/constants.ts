/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import {
  FieldMappingPropertyMap,
  GetFieldMappingViewResponse,
} from '../../../../../../server/models/interfaces';

export const STATUS_ICON_PROPS = {
  unmapped: { type: 'alert', color: 'danger' },
  mapped: { type: 'checkInCircleFilled', color: 'success' },
};

export const EMPTY_FIELD_MAPPINGS_VIEW: GetFieldMappingViewResponse = {
  properties: {},
  unmapped_field_aliases: [],
  unmapped_index_fields: [],
};

export const EMPTY_FIELD_MAPPINGS: FieldMappingPropertyMap = {
  properties: {},
};
