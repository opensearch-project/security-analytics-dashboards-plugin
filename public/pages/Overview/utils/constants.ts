/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

export const summaryGroupByOptions = [
  { text: 'All findings', value: 'all_findings' },
  { text: 'Log type', value: 'log_type' },
];

export const moreLink = 'https://opensearch.org/docs/latest/security-analytics/';
