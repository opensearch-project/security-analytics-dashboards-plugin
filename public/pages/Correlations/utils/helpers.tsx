/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { EuiBasicTableColumn, EuiBadge, EuiToolTip, EuiSmallButtonIcon, EuiLink } from '@elastic/eui';
import { CorrelationRule, CorrelationRuleQuery, CorrelationRuleTableItem } from '../../../../types';
import { Search } from '@opensearch-project/oui/src/eui_components/basic_table';
import { formatRuleType, getLogTypeFilterOptions } from '../../../utils/helpers';

export const getCorrelationRulesTableColumns = (
  onRuleNameClick: (rule: CorrelationRule) => void,
  _refreshRules: (ruleItem: CorrelationRule) => void
): EuiBasicTableColumn<CorrelationRuleTableItem>[] => {
  return [
    {
      field: 'name',
      name: 'Name',
      sortable: true,
      truncateText: true,
      render: (name: string, ruleItem: CorrelationRule) => (
        <EuiLink onClick={() => onRuleNameClick(ruleItem)}>{name}</EuiLink>
      ),
    },
    {
      name: 'Log types',
      field: 'logTypes',
      render: (logTypes: string, ruleItem: CorrelationRule) => {
        const badges = [
          ...new Set(ruleItem.queries?.map((query) => formatRuleType(query.logType))),
        ];
        return (
          <>
            {badges.map((badge) => (
              <EuiBadge color="hollow">{badge}</EuiBadge>
            ))}
          </>
        );
      },
    },
    {
      field: 'queries',
      name: 'Queries',
      align: 'right',
      render: (queries: CorrelationRuleQuery[], ruleItem: CorrelationRule) => {
        return queries.length;
      },
      width: '10%',
    },
    {
      name: 'Actions',
      field: '',
      actions: [
        {
          render: (ruleItem: CorrelationRule) => (
            <EuiToolTip content={'Delete'}>
              <EuiSmallButtonIcon
                aria-label={'Delete correlation rule'}
                data-test-subj={`view-details-icon`}
                iconType={'trash'}
                color="danger"
                onClick={() => _refreshRules(ruleItem)}
              />
            </EuiToolTip>
          ),
        },
      ],
    },
  ];
};

export const getCorrelationRulesTableSearchConfig = (): Search => {
  return {
    box: {
      placeholder: 'Search by rule name, log type',
      schema: true,
    },
    filters: [
      {
        type: 'field_value_selection',
        field: 'logTypes',
        name: 'Log Types',
        multiSelect: 'or',
        options: getLogTypeFilterOptions(),
      },
    ],
  };
};
