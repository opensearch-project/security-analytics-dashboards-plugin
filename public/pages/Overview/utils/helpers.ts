/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { euiPaletteColorBlind, euiPaletteForStatus } from '@elastic/eui';
import { TopLevelSpec } from 'vega-lite';
import { SummaryData } from '../components/Widgets/Summary';
import dateMath from '@elastic/datemath';
import _ from 'lodash';
import { DEFAULT_DATE_RANGE } from '../../../utils/constants';
import { severityOptions } from '../../Alerts/utils/constants';

export interface TimeUnit {
  unit: string;
  step: number;
}

export interface TimeUnits {
  timeUnit: TimeUnit;
  dateFormat: string;
}

export type DateOpts = {
  timeUnit: TimeUnit;
  dateFormat: string;
  domain?: number[];
};

/**
 * Legend selection config for the chart layer
 */
export const legendSelectionCfg = {
  selection: {
    series: {
      type: 'multi',
      encodings: ['color'],
      on: 'click',
      bind: 'legend',
    },
  },
  encoding: {
    opacity: {
      condition: { selection: 'series', value: 1 },
      value: 0.2,
    },
  },
};

export const defaultTimeUnit = {
  unit: 'yearmonthdatehoursminutes',
  step: 1,
};

export const defaultDateFormat = '%Y-%m-%d %H:%M';

// euiColorDanger: #BD271E
export const alertsDefaultColor = '#BD271E';

export const parseDateString = (dateString: string): number => {
  const date = dateMath.parse(dateString);
  return date ? date.toDate().getTime() : new Date().getTime();
};

export const defaultScaleDomain = [
  parseDateString(DEFAULT_DATE_RANGE.start),
  parseDateString(DEFAULT_DATE_RANGE.end),
];

export const getYAxis = (field: string, title: string, axisGrid: boolean = true) => ({
  aggregate: 'sum',
  field: field,
  type: 'quantitative',
  title: title,
  axis: { grid: axisGrid },
});

export const getXAxis = (dateOpts: DateOpts, opts: any = {}) =>
  _.defaultsDeep(opts, {
    timeUnit: dateOpts.timeUnit,
    field: 'time',
    title: '',
    type: 'temporal',
    axis: { grid: false, format: dateOpts.dateFormat },
    scale: {
      domain: dateOpts.domain,
    },
  });

export const getTimeTooltip = (dateOpts: DateOpts) => ({
  timeUnit: dateOpts.timeUnit,
  field: 'time',
  type: 'temporal',
  title: 'Time',
  format: '%Y-%m-%d %H:%M:%S',
});

export function getVisualizationSpec(description: string, data: any, layers: any[]): TopLevelSpec {
  return {
    config: {
      view: { stroke: 'transparent' },
      legend: {
        labelColor: '#343741',
        titleColor: '#1a1c21',
        labelFontSize: 14,
        titleFontWeight: 600,
        titleLineHeight: 21,
        titleFontSize: 14,
        titlePadding: 10,
        rowPadding: 6,
        labelFont:
          '"Inter UI", -apple-system, BlinkMacSystemFont, "Segoe UI", Helvetica, Arial, sans-serif, "Apple Color Emoji", "Segoe UI Emoji", "Segoe UI Symbol"',
      },
    },
    $schema: 'https://vega.github.io/schema/vega-lite/v5.json',
    description: description,
    data: {
      values: data,
    },
    layer: layers,
  };
}

export function getOverviewVisualizationSpec(
  visualizationData: SummaryData[],
  groupBy: string,
  dateOpts: DateOpts = {
    timeUnit: defaultTimeUnit,
    dateFormat: defaultDateFormat,
    domain: defaultScaleDomain,
  }
): TopLevelSpec {
  const findingsEncoding: { [x: string]: any } = {
    x: getXAxis(dateOpts),
    y: getYAxis('finding', 'Count'),
    tooltip: [getYAxis('finding', 'Findings'), getTimeTooltip(dateOpts)],
    color: {
      field: 'fieldType',
      title: '',
      legend: {
        values: ['Active alerts', 'Findings'],
      },
      scale: {
        domain: ['Active alerts', 'Findings'],
        range: [alertsDefaultColor, euiPaletteColorBlind()[1]],
      },
    },
  };

  let barLayer = {
    mark: {
      type: 'bar',
      clip: true,
    },
    transform: [{ calculate: "datum.alert == 1 ? 'Active alerts' : 'Findings'", as: 'fieldType' }],
    encoding: findingsEncoding,
  };

  if (groupBy === 'logType') {
    findingsEncoding['color'] = {
      field: 'logType',
      type: 'nominal',
      title: 'Log type',
      scale: {
        range: euiPaletteColorBlind(),
      },
    };

    findingsEncoding['tooltip'].push({
      field: groupBy,
      title: groupBy === 'logType' ? 'Log type' : 'Rule severity',
    });

    barLayer = addInteractiveLegends(barLayer);
  }

  return getVisualizationSpec(
    'Plot showing average data with raw values in the background.',
    visualizationData,
    [
      barLayer,
      {
        mark: {
          type: 'line',
          clip: true,
          interpolate: 'monotone',
          color: alertsDefaultColor,
          point: {
            filled: false,
            fill: 'white',
            color: alertsDefaultColor,
            size: 50,
          },
        },
        encoding: {
          x: getXAxis(dateOpts, {
            band: 0.5,
          }),
          y: getYAxis('alert', 'Count'),
          tooltip: [getYAxis('alert', 'Alerts'), getTimeTooltip(dateOpts)],
        },
      },
    ]
  );
}

export function getFindingsVisualizationSpec(
  visualizationData: any[],
  groupBy: string,
  dateOpts: DateOpts = {
    timeUnit: defaultTimeUnit,
    dateFormat: defaultDateFormat,
    domain: defaultScaleDomain,
  }
) {
  const severities = ['info', 'low', 'medium', 'high', 'critical'];
  const isGroupedByLogType = groupBy === 'logType';
  const logTitle = 'Log type';
  const severityTitle = 'Rule severity';
  const title = isGroupedByLogType ? logTitle : severityTitle;
  return getVisualizationSpec('Findings data overview', visualizationData, [
    addInteractiveLegends({
      mark: {
        type: 'bar',
        clip: true,
      },
      encoding: {
        tooltip: [
          getYAxis('finding', 'Findings'),
          getTimeTooltip(dateOpts),
          {
            field: groupBy,
            title: title,
          },
        ],
        x: getXAxis(dateOpts),
        y: getYAxis('finding', 'Count'),
        color: {
          field: groupBy,
          title: title,
          scale: {
            domain: isGroupedByLogType ? undefined : severities,
            range: groupBy === 'logType' ? euiPaletteColorBlind() : euiPaletteForStatus(5),
          },
        },
      },
    }),
  ]);
}

export function getAlertsVisualizationSpec(
  visualizationData: any[],
  groupBy: string,
  dateOpts: DateOpts = {
    timeUnit: defaultTimeUnit,
    dateFormat: defaultDateFormat,
    domain: defaultScaleDomain,
  }
) {
  const isGroupedByStatus = groupBy === 'status';
  let severities = severityOptions.map((severity) => severity.text);
  severities.reverse().pop();

  let states = ['ACTIVE', 'ACKNOWLEDGED'];
  const statusColors = {
    euiColorVis9: '#E7664C',
    euiColorVis6: '#B9A888',
  };

  const statusTitle = 'Alert status';
  const severityTitle = 'Alert severity';
  const title = isGroupedByStatus ? statusTitle : severityTitle;
  return getVisualizationSpec('Alerts data overview', visualizationData, [
    addInteractiveLegends({
      mark: {
        type: 'bar',
        clip: true,
      },
      encoding: {
        tooltip: [
          getYAxis('alert', 'Alerts'),
          getTimeTooltip(dateOpts),
          {
            field: groupBy,
            title: title,
          },
        ],
        x: getXAxis(dateOpts),
        y: getYAxis('alert', 'Count'),
        color: {
          field: groupBy,
          title: title,
          scale: {
            domain: isGroupedByStatus ? states : severities,
            range: isGroupedByStatus ? Object.values(statusColors) : euiPaletteForStatus(5),
          },
        },
      },
    }),
  ]);
}

export function getTopRulesVisualizationSpec(visualizationData: any[]) {
  return getVisualizationSpec('Most frequent detection rules', visualizationData, [
    {
      mark: { type: 'arc', innerRadius: 90 },
      transform: [
        {
          joinaggregate: [
            {
              op: 'sum',
              field: 'count',
              as: 'total',
            },
          ],
        },
        {
          calculate: 'datum.count/datum.total',
          as: 'percentage',
        },
      ],
      encoding: {
        tooltip: [
          {
            field: 'percentage',
            title: 'Percentage',
            type: 'quantitative',
            format: '2.0%',
          },
          {
            field: 'ruleName',
            type: 'nominal',
            title: 'Rule',
          },
          getYAxis('count', 'Count'),
        ],
        theta: getYAxis('count', ''),
        color: {
          field: 'ruleName',
          type: 'nominal',
          title: 'Rule name',
          scale: {
            range: euiPaletteColorBlind(),
          },
        },
      },
    },
  ]);
}

export function getTimeWithMinPrecision(time: number | string) {
  const date = new Date(time);
  date.setSeconds(0);
  date.setMilliseconds(0);

  return date.getTime();
}

/**
 * Returns timeUnit based on how big time diff is between start and end dates
 * @param start Chart start time
 * @param end Chart end time
 * @param [defaultUnit = 'yearmonthdatehoursminutes'] Default timeUnit
 */
export function getChartTimeUnit(
  start: string,
  end: string,
  defaultUnit: string = 'yearmonthdatehoursminutes'
): TimeUnits {
  const startMoment = dateMath.parse(start);
  const endMoment = dateMath.parse(end);
  let timeUnit: string = defaultUnit;
  let dateFormat: string = '%Y-%m-%d %H:%M';

  if (!startMoment || !endMoment)
    return {
      timeUnit: { unit: timeUnit, step: 1 },
      dateFormat,
    };

  try {
    const milliseconds = endMoment.diff(startMoment);

    const second = 1001; // set 1ms as a threshold since moment can make a mistake in 1 ms when calculating start and end datetime
    const minute = second * 60;
    const hour = minute * 60;
    const day = hour * 24;
    const month = day * 30;
    const year = month * 12;

    if (milliseconds <= minute) {
      timeUnit = 'yearmonthdatehoursminutesseconds';
      dateFormat = '%Y-%m-%d %H:%M:%S';
    } else if (milliseconds <= hour) {
      timeUnit = 'yearmonthdatehoursminutes';
      dateFormat = '%Y-%m-%d %H:%M';
    } else if (milliseconds <= day * 2) {
      timeUnit = 'yearmonthdatehours';
      dateFormat = '%Y-%m-%d %H:%M';
    } else if (milliseconds <= month * 6) {
      timeUnit = 'yearmonthdate';
      dateFormat = '%Y-%m-%d';
    } else if (milliseconds <= year * 2) {
      timeUnit = 'yearmonth';
      dateFormat = '%Y-%m';
    } else if (milliseconds <= year * 6) {
      timeUnit = 'yearquarter';
      dateFormat = '%Y';
    } else {
      timeUnit = 'year';
      dateFormat = '%Y';
    }
  } catch (e) {
    console.error(`Time diff can't be calculated for dates: ${start} and ${end}`);
  }
  return {
    timeUnit: _.assign(
      {
        step: 1,
      },
      { unit: timeUnit }
    ),
    dateFormat,
  };
}

/**
 * Adds interactive legends to the chart layer
 * @param layer
 */
export const addInteractiveLegends = (layer: any) => _.defaultsDeep(layer, legendSelectionCfg);

export const getDomainRange = (
  range: string[] = [DEFAULT_DATE_RANGE.start, DEFAULT_DATE_RANGE.end],
  timeUnit?: string
): number[] => {
  const start: number = parseDateString(range[0] || DEFAULT_DATE_RANGE.start);
  let rangeEnd = range[1] || DEFAULT_DATE_RANGE.end;
  if (timeUnit) {
    const timeUnitSize = timeUnit.match(/.*(seconds|minutes|hours|date|month|year)$/);
    if (timeUnitSize && timeUnitSize[1]) {
      rangeEnd = `now+1${timeUnitSize[1][0]}`;
    }
  }
  const end: number = parseDateString(rangeEnd);
  return [start, end];
};
