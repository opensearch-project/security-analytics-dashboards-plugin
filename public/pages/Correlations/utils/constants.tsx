/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import React from 'react';
import { CorrelationGraphData } from '../../../../types';
import { ruleSeverity, ruleTypes } from '../../Rules/utils/constants';
import { FilterItem } from '../components/FilterGroup';
import { EuiIcon, EuiTitle } from '@elastic/eui';
import { logTypeCategories, logTypesByCategories } from '../../../utils/constants';
import { getLogTypeLabel } from '../../LogTypes/utils/helpers';

export const graphRenderOptions = {
  nodes: {
    shape: 'dot',
  },
  edges: {
    arrows: {
      to: {
        enabled: false,
      },
    },
    color: '#000000',
    chosen: true,
  },
  layout: {
    hierarchical: false,
    randomSeed: 2222,
    improvedLayout: false,
  },
  autoResize: true,
  height: '800px',
  width: '100%',
  physics: {
    stabilization: {
      fit: true,
      iterations: 1000,
    },
  },
  interaction: {
    zoomView: true,
    zoomSpeed: 0.2,
    dragView: true,
    dragNodes: false,
    multiselect: true,
    tooltipDelay: 50,
    hover: true,
  },
};

export const getDefaultLogTypeFilterItemOptions: () => FilterItem[] = () => {
  const options: FilterItem[] = [];
  logTypeCategories.forEach((category) => {
    const logTypes = logTypesByCategories[category];
    options.push({
      name: (
        <EuiTitle size="xxs">
          <h4>{category}</h4>
        </EuiTitle>
      ),
      id: category,
      checked: 'on',
      childOptionIds: new Set(logTypes.map(({ name }) => name)),
      visible: true,
    });

    logTypes.forEach(({ name }) => {
      options.push({
        name: getLogTypeLabel(name),
        id: name,
        checked: 'on',
        visible: true,
      });
    });
  });

  return options;
};

export const defaultSeverityFilterItemOptions: FilterItem[] = Object.values(ruleSeverity).map(
  (sev) => {
    return {
      name: (
        <p>
          <EuiIcon type={'dot'} color={sev.color.background} /> {`${sev.name}`}
        </p>
      ),
      id: sev.value,
      checked: 'on',
      visible: true,
    };
  }
);

export const emptyGraphData: CorrelationGraphData = {
  graph: {
    nodes: [],
    edges: [],
  },
  events: {},
};

export const getLabelFromLogType = (logType: string) => {
  return ruleTypes.find((ruleType) => ruleType.value === logType)?.label || logType;
};

export const getSeverityLabel = (sev: string) => {
  return ruleSeverity.find((severity) => severity.value === sev)?.name || '';
};

export const getSeverityColor = (sev: string) => {
  return (
    ruleSeverity.find((severity) => severity.value === sev.toLowerCase())?.color || {
      background: 'white',
      text: 'black',
    }
  );
};

export function getNodeSize(severity: string) {
  switch (severity) {
    case 'critical':
      return 40;
    case 'high':
      return 32;
    case 'medium':
      return 24;
    case 'low':
      return 17;
    case 'informational':
    default:
      return 10;
  }
}
