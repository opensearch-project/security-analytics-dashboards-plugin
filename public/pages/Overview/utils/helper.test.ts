import { getChartTimeUnit, getDateFormatByTimeUnit } from './helpers';
import { TimeUnitsMap } from './constants';
import moment from 'moment';

describe('helper utilities spec', () => {
  describe('tests getDateFormatByTimeUnit function', () => {
    const yearFormat = '%Y-%m-%d';
    const dayFormat = '%H:%M:%S';
    const fullFormat = '%Y-%m-%d %H:%M';
    const hoursAgo = moment().subtract(15, 'hours');

    const timeFormats: {
      [key: string]: string;
    } = {
      'now-15m': dayFormat,
      'now-15h': hoursAgo.date() === moment().date() ? dayFormat : fullFormat,
      'now-15d': fullFormat,
      'now-2M': yearFormat,
      'now-2y': fullFormat,
    };

    it(` - function should return default format ${fullFormat} if dates are not valid`, () => {
      expect(getDateFormatByTimeUnit('', '')).toBe(fullFormat);
    });

    for (const [start, format] of Object.entries(timeFormats)) {
      it(` - function should return ${format} if start date is ${start}`, () => {
        expect(getDateFormatByTimeUnit(start, 'now')).toBe(format);
      });
    }
  });

  describe('tests getChartTimeUnit function', () => {
    const defaultTimeUnit = 'yearmonthdatehoursminutes';
    it(' - function should return default timeUnit if fn params are invalid', () => {
      expect(getChartTimeUnit('', '')).toBe(defaultTimeUnit);
    });

    it(' - function should return default timeUnit if one is passed as param', () => {
      const defaultFormat = 'yearmonthdate';
      expect(getChartTimeUnit('', '', defaultFormat)).toBe(defaultFormat);
    });

    const timeUnits: {
      [key: string]: string;
    } = {
      minutes: 'now-15m',
      hours: 'now-15h',
      days: 'now-15d',
      weeks: 'now-15w',
      months: 'now-5M',
      years: 'now-15y',
    };

    for (const [unit, start] of Object.entries(timeUnits)) {
      it(` - function should return ${TimeUnitsMap[unit]} if unit is ${unit}`, () => {
        expect(getChartTimeUnit(start, 'now')).toBe(TimeUnitsMap[unit]);
      });
    }
  });
});
