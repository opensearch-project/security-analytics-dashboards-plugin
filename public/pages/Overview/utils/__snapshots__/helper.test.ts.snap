// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`helper utilities spec tests getAlertsVisualizationSpec function  - snapshot test: should match alerts spec 1`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "config": Object {
    "legend": Object {
      "labelColor": "#343741",
      "labelFont": "\\"Inter UI\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
      "labelFontSize": 14,
      "rowPadding": 6,
      "titleColor": "#1a1c21",
      "titleFontSize": 14,
      "titleFontWeight": 600,
      "titleLineHeight": 21,
      "titlePadding": 10,
    },
    "view": Object {
      "stroke": "transparent",
    },
  },
  "data": Object {
    "values": Array [],
  },
  "description": "Alerts data overview",
  "layer": Array [
    Object {
      "encoding": Object {
        "color": Object {
          "field": "",
          "scale": Object {
            "domain": Array [
              "5 (Lowest)",
              "4 (Low)",
              "3 (Medium)",
              "2 (High)",
              "1 (Highest)",
            ],
            "range": Array [
              "#209280",
              "#54b399",
              "#d6bf57",
              "#e7664c",
              "#cc5642",
            ],
          },
          "title": "Alert severity",
        },
        "opacity": Object {
          "condition": Object {
            "selection": "series",
            "value": 1,
          },
          "value": 0.2,
        },
        "tooltip": Array [
          Object {
            "aggregate": "sum",
            "axis": Object {
              "grid": true,
            },
            "field": "alert",
            "title": "Alerts",
            "type": "quantitative",
          },
          Object {
            "field": "time",
            "format": "%Y-%m-%d %H:%M:%S",
            "timeUnit": Object {
              "step": 1,
              "unit": "yearmonthdatehoursminutes",
            },
            "title": "Time",
            "type": "temporal",
          },
          Object {
            "field": "",
            "title": "Alert severity",
          },
        ],
        "x": Object {
          "axis": Object {
            "format": "%Y-%m-%d %H:%M",
            "grid": false,
          },
          "field": "time",
          "scale": Object {
            "domain": undefined,
          },
          "timeUnit": Object {
            "step": 1,
            "unit": "yearmonthdatehoursminutes",
          },
          "title": "",
          "type": "temporal",
        },
        "y": Object {
          "aggregate": "sum",
          "axis": Object {
            "grid": true,
          },
          "field": "alert",
          "title": "Count",
          "type": "quantitative",
        },
      },
      "mark": Object {
        "clip": true,
        "type": "bar",
      },
      "selection": Object {
        "series": Object {
          "bind": "legend",
          "encodings": Array [
            "color",
          ],
          "on": "click",
          "type": "multi",
        },
      },
    },
  ],
}
`;

exports[`helper utilities spec tests getFindingsVisualizationSpec function  - snapshot test: should match findings spec 1`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "config": Object {
    "legend": Object {
      "labelColor": "#343741",
      "labelFont": "\\"Inter UI\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
      "labelFontSize": 14,
      "rowPadding": 6,
      "titleColor": "#1a1c21",
      "titleFontSize": 14,
      "titleFontWeight": 600,
      "titleLineHeight": 21,
      "titlePadding": 10,
    },
    "view": Object {
      "stroke": "transparent",
    },
  },
  "data": Object {
    "values": Array [],
  },
  "description": "Findings data overview",
  "layer": Array [
    Object {
      "encoding": Object {
        "color": Object {
          "field": "",
          "scale": Object {
            "domain": Array [
              "info",
              "low",
              "medium",
              "high",
              "critical",
            ],
            "range": Array [
              "#209280",
              "#54b399",
              "#d6bf57",
              "#e7664c",
              "#cc5642",
            ],
          },
          "title": "Rule severity",
        },
        "opacity": Object {
          "condition": Object {
            "selection": "series",
            "value": 1,
          },
          "value": 0.2,
        },
        "tooltip": Array [
          Object {
            "aggregate": "sum",
            "axis": Object {
              "grid": true,
            },
            "field": "finding",
            "title": "Findings",
            "type": "quantitative",
          },
          Object {
            "field": "time",
            "format": "%Y-%m-%d %H:%M:%S",
            "timeUnit": Object {
              "step": 1,
              "unit": "yearmonthdatehoursminutes",
            },
            "title": "Time",
            "type": "temporal",
          },
          Object {
            "field": "",
            "title": "Rule severity",
          },
        ],
        "x": Object {
          "axis": Object {
            "format": "%Y-%m-%d %H:%M",
            "grid": false,
          },
          "field": "time",
          "scale": Object {
            "domain": undefined,
          },
          "timeUnit": Object {
            "step": 1,
            "unit": "yearmonthdatehoursminutes",
          },
          "title": "",
          "type": "temporal",
        },
        "y": Object {
          "aggregate": "sum",
          "axis": Object {
            "grid": true,
          },
          "field": "finding",
          "title": "Count",
          "type": "quantitative",
        },
      },
      "mark": Object {
        "clip": true,
        "type": "bar",
      },
      "selection": Object {
        "series": Object {
          "bind": "legend",
          "encodings": Array [
            "color",
          ],
          "on": "click",
          "type": "multi",
        },
      },
    },
  ],
}
`;

exports[`helper utilities spec tests getOverviewVisualizationSpec function  - snapshot test: should match overview spec 1`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "config": Object {
    "legend": Object {
      "labelColor": "#343741",
      "labelFont": "\\"Inter UI\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
      "labelFontSize": 14,
      "rowPadding": 6,
      "titleColor": "#1a1c21",
      "titleFontSize": 14,
      "titleFontWeight": 600,
      "titleLineHeight": 21,
      "titlePadding": 10,
    },
    "view": Object {
      "stroke": "transparent",
    },
  },
  "data": Object {
    "values": Array [],
  },
  "description": "Plot showing average data with raw values in the background.",
  "layer": Array [
    Object {
      "encoding": Object {
        "color": Object {
          "field": "fieldType",
          "legend": Object {
            "values": Array [
              "Active alerts",
              "Findings",
            ],
          },
          "scale": Object {
            "domain": Array [
              "Active alerts",
              "Findings",
            ],
            "range": Array [
              "#BD271E",
              "#6092C0",
            ],
          },
          "title": "",
        },
        "tooltip": Array [
          Object {
            "aggregate": "sum",
            "axis": Object {
              "grid": true,
            },
            "field": "finding",
            "title": "Findings",
            "type": "quantitative",
          },
          Object {
            "field": "time",
            "format": "%Y-%m-%d %H:%M:%S",
            "timeUnit": Object {
              "step": 1,
              "unit": "yearmonthdatehoursminutes",
            },
            "title": "Time",
            "type": "temporal",
          },
        ],
        "x": Object {
          "axis": Object {
            "format": "%Y-%m-%d %H:%M",
            "grid": false,
          },
          "field": "time",
          "scale": Object {
            "domain": undefined,
          },
          "timeUnit": Object {
            "step": 1,
            "unit": "yearmonthdatehoursminutes",
          },
          "title": "",
          "type": "temporal",
        },
        "y": Object {
          "aggregate": "sum",
          "axis": Object {
            "grid": true,
          },
          "field": "finding",
          "title": "Count",
          "type": "quantitative",
        },
      },
      "mark": Object {
        "clip": true,
        "type": "bar",
      },
      "transform": Array [
        Object {
          "as": "fieldType",
          "calculate": "datum.alert == 1 ? 'Active alerts' : 'Findings'",
        },
      ],
    },
    Object {
      "encoding": Object {
        "tooltip": Array [
          Object {
            "aggregate": "sum",
            "axis": Object {
              "grid": true,
            },
            "field": "alert",
            "title": "Alerts",
            "type": "quantitative",
          },
          Object {
            "field": "time",
            "format": "%Y-%m-%d %H:%M:%S",
            "timeUnit": Object {
              "step": 1,
              "unit": "yearmonthdatehoursminutes",
            },
            "title": "Time",
            "type": "temporal",
          },
        ],
        "x": Object {
          "axis": Object {
            "format": "%Y-%m-%d %H:%M",
            "grid": false,
          },
          "band": 0.5,
          "field": "time",
          "scale": Object {
            "domain": undefined,
          },
          "timeUnit": Object {
            "step": 1,
            "unit": "yearmonthdatehoursminutes",
          },
          "title": "",
          "type": "temporal",
        },
        "y": Object {
          "aggregate": "sum",
          "axis": Object {
            "grid": true,
          },
          "field": "alert",
          "title": "Count",
          "type": "quantitative",
        },
      },
      "mark": Object {
        "clip": true,
        "color": "#BD271E",
        "interpolate": "monotone",
        "point": Object {
          "color": "#BD271E",
          "fill": "white",
          "filled": false,
          "size": 50,
        },
        "type": "line",
      },
    },
  ],
}
`;

exports[`helper utilities spec tests getVisualizationSpec function  - snapshot test: should match visualization spec 1`] = `
Object {
  "$schema": "https://vega.github.io/schema/vega-lite/v5.json",
  "config": Object {
    "legend": Object {
      "labelColor": "#343741",
      "labelFont": "\\"Inter UI\\", -apple-system, BlinkMacSystemFont, \\"Segoe UI\\", Helvetica, Arial, sans-serif, \\"Apple Color Emoji\\", \\"Segoe UI Emoji\\", \\"Segoe UI Symbol\\"",
      "labelFontSize": 14,
      "rowPadding": 6,
      "titleColor": "#1a1c21",
      "titleFontSize": 14,
      "titleFontWeight": 600,
      "titleLineHeight": 21,
      "titlePadding": 10,
    },
    "view": Object {
      "stroke": "transparent",
    },
  },
  "data": Object {
    "values": Array [
      Object {
        "xField": 1,
        "yField": 1,
      },
    ],
  },
  "description": "Visualization",
  "layer": Array [
    Object {
      "encoding": Object {
        "opacity": Object {
          "condition": Object {
            "selection": "series",
            "value": 1,
          },
          "value": 0.2,
        },
      },
      "selection": Object {
        "series": Object {
          "bind": "legend",
          "encodings": Array [
            "color",
          ],
          "on": "click",
          "type": "multi",
        },
      },
      "x": Object {
        "field": "xField",
      },
      "y": Object {
        "field": "yField",
      },
    },
  ],
}
`;
